src/lib/device-online-state.ts

```diff
@@ -0,0 +1,337 @@
+		// return early if the feature isn't active...
+		if (API_HEARTBEAT_STATE_ACTIVE !== 1) {
+			return;
+		}
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch(err => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				return this.consume();
+			});
+	}
+
+	private updateDeviceModel(uuid: string, newState: DeviceOnlineStates) {
+		// patch the api_heartbeat_state value to the new state...
+		const body = {
+			api_heartbeat_state: newState,
+		};
+
+		return resinApi
+			.patch({
+				resource: 'device',
+				options: {
+					$filter: {
+						uuid,
+						$not: body,
+					},
+				},
+				body,
+				passthrough: {
+					req: root,
+				},
+			})
+			.return(true)
+			.catch(err => {
+				captureException(
+					err,
+					'DeviceStateManager: Error updating the API with the device new state.',
+				);
+
+				return false;
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		return this.updateDeviceModel(uuid, DeviceOnlineStates.Online);
+	}
+
+	private consume() {
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessageAsync({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: DeviceOnlineStateManager.RSMQ_READ_TIMEOUT, // prevent other consumers seeing the same message (if any) preventing multiple API agents from processing it...
+			})
+			.then(msg => {
+				if ('id' in msg) {
+					const { id, message } = msg;
+
+					return Bluebird.try(() => {
+						const { uuid, nextState } = JSON.parse(message) as {
+							uuid: string;
+							nextState: DeviceOnlineStates;
+						};
+
+						// raise and event for the state change...
+						switch (nextState) {
+							case DeviceOnlineStates.Timeout:
+								this.scheduleChangeOfStateForDevice(
+									uuid,
+									DeviceOnlineStates.Timeout,
+									DeviceOnlineStates.Offline,
+									API_HEARTBEAT_STATE_TIMEOUT_SECONDS, // put the device into a timeout state if it misses it's scheduled heartbeat window... then mark as offline
+								);
+								return this.updateDeviceModel(uuid, DeviceOnlineStates.Timeout);
+							case DeviceOnlineStates.Offline:
+								return this.updateDeviceModel(uuid, DeviceOnlineStates.Offline);
+							default:
+								throw new Error(
+									`An unexpected value was encountered for the target device state: ${nextState}`,
+								);
+						}
+					})
+						.then(() =>
+							this.rsmq.deleteMessageAsync({
+								qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+								id,
+							}),
+						)
+						.catch((err: Error) => captureException(err));
```

Comment:
Can you add a message to this `captureException` please, so it's easier to trace in sentry